src/absorbable.rs

```diff
@@ -0,0 +1,330 @@
+    fn batch_to_sponge_field_elements(batch: &[Self]) -> Vec<F>
+    where
+        Self: Sized,
+    {
+        let mut output = Vec::new();
+        for absorbable in batch {
+            output.append(&mut absorbable.to_sponge_field_elements());
+        }
+
+        output
+    }
+
+    /// Specifies the conversion into a list of field elements for a batch along with its length information.
+    fn batch_to_sponge_field_elements_with_length(batch: &[Self]) -> Vec<F>
+    where
+        Self: Sized,
+    {
+        let mut output = Vec::new();
+        output.append(&mut <u64 as Absorb<F>>::to_sponge_field_elements(
+            &(batch.len() as u64),
+        ));
+        for absorbable in batch {
+            output.append(&mut absorbable.to_sponge_field_elements());
+        }
+
+        output
+    }
+}
+
+/// An extension to `Absorb` interface that allows an option for sponge to absorb length information.
+/// This extended interface should be used in cases when the length of `self` is list-like and
+pub trait AbsorbWithLength<F: PrimeField>: Absorb<F> {
+    /// Converts the object and its length information into a list of bytes that
+    /// can be absorbed by a `CryptographicSponge`.
+    fn to_sponge_bytes_with_length(&self) -> Vec<u8>;
+
+    /// Converts the object and its length information into field elements that can be absorbed by a `CryptographicSponge`.
+    fn to_sponge_field_elements_with_length(&self) -> Vec<F>;
+}
+
+impl<F: PrimeField> Absorb<F> for u8 {
+    fn to_sponge_bytes(&self) -> Vec<u8> {
+        vec![*self]
+    }
+
+    fn to_sponge_field_elements(&self) -> Vec<F> {
+        vec![F::from(*self)]
+    }
+
+    fn batch_to_sponge_bytes(batch: &[Self]) -> Vec<u8> {
+        batch.to_vec()
+    }
+
+    fn batch_to_sponge_field_elements(batch: &[Self]) -> Vec<F> {
+        let mut bytes = (batch.len() as u64).to_le_bytes().to_vec();
+        bytes.extend_from_slice(batch);
+        bytes.to_field_elements().unwrap()
+    }
+}
+
+impl<F: PrimeField> Absorb<F> for bool {
+    fn to_sponge_bytes(&self) -> Vec<u8> {
+        vec![(*self as u8)]
+    }
+
+    fn to_sponge_field_elements(&self) -> Vec<F> {
+        vec![F::from(*self)]
+    }
+}
+
+macro_rules! impl_absorbable_field {
+    ($field:ident, $params:ident) => {
+        impl<P: $params> Absorb<$field<P>> for $field<P> {
+            fn to_sponge_bytes(&self) -> Vec<u8> {
+                to_bytes![self].unwrap()
+            }
+
+            fn to_sponge_field_elements(&self) -> Vec<$field<P>> {
+                vec![*self]
+            }
+
+            fn batch_to_sponge_field_elements(batch: &[Self]) -> Vec<$field<P>> {
+                batch.to_vec()
+            }
+        }
+    };
+}
+
+impl_absorbable_field!(Fp256, Fp256Parameters);
+impl_absorbable_field!(Fp320, Fp320Parameters);
+impl_absorbable_field!(Fp384, Fp384Parameters);
+impl_absorbable_field!(Fp768, Fp768Parameters);
+impl_absorbable_field!(Fp832, Fp832Parameters);
```

Comment:
So for now is there a plan to implement Absorb for non-native field? Current implementation seems to require that field can only be native (i.e. Fp256 can only be absorbed by Sponge<Fp256>)